design/src/rfcs/rfc0023_request_id.md

```diff
@@ -0,0 +1,105 @@
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+Terminology
+-----------
+
+- **RequestID**: a service-wide unique identifier
+- **UUID**: a universally unique identifier
+
+<!-- Insert a short paragraph explaining, at a high level, what this RFC is for -->
+RequestID is an element that uniquely identifies a client request. RequestID is used by services to map all logs, events and
+specific data to a single operation. This RFC discusses whether and how smithy-rs can make that value available to customers.
+
+Services use a RequestID to collect logs related to the same request and see its flow through the various operations,
+help clients debug requests by sharing this value and, in some cases, use this value to perform their business logic. RequestID is unique across a service.
+
+This value for the purposes above must be set by the service. Clients could send any value not conforming to the valid, chosen format,
+avoid sending one altogether or exploit how some service is using that value
+(like in [1](https://en.wikipedia.org/wiki/Shellshock_(software_bug)) and
+[2](https://cwiki.apache.org/confluence/display/WW/S2-045)).
+Having the client send the value bring the following challenges:
+* The client could repeatedly send the same RequestID
+* The client could send no RequestID
+* The client could send a malformed or malicious RequestID
+
+To minimise the attack surface and provide a uniform experience to customers, servers must generate the value.
+Services are free to read the ID sent by clients in HTTP headers.
+
+RequestIDs are not to be used by multiple services, but only within a service.
+
+<!-- Explain how users will use this new feature and, if necessary, how this compares to the current user experience -->
+The user experience if this RFC is implemented
+----------------------------------------------
+
+The proposal is to implement a `RequestId` type and make it available to middleware and business logic handlers, through [FromParts](../server/from-parts.md) and as a `Service`.
+
+1. Implementing `FromParts` for `Extension<RequestId>` gives customers the ability to write their handlers:
+
+```rust
+pub async fn handler(
+    input: input::Input,
+    request_id: Extension<RequestId>,
+) -> ...
+```
+
+`RequestId` will be injected into the extensions by a layer.
+This layer can also be used to open a span that will log the request ID: subsequent logs will be in the scope of that span.
+
+2. RequestId format:
+
+Common formats for RequestIDs are:
+
+* UUID: a random string, represented in hex, of 128 bits from IETF RFC 4122: `7c038a43-e499-4162-8e70-2d4d38595930`
+* The hash of a sequence such as `date+thread+server`: `734678902ea938783a7200d7b2c0b487`
+* A verbose description: `current_ms+hostname+increasing_id`
+
+For privacy reasons, any format that provides service details should be avoided. A random string is preferred.
+The proposed format is to use UUID, version 4.
```

Comment:
Have we considered v7?
It is pretty much equivalent in terms of uniqueness guarantees, but it provides better performance if a service decides to store them.

Reply:
I'll take a look, but it seems to me like the RFC is still in draft for v7

Reply:
Yes, it is still in draft. 
Do we consider the format to be part of the public API of the request ID (i.e. is it exposed to users)? Or do we simply want to users to treat is as an opaque new type that can be converted into a `String` (therefore allowing us to evolve the internal implementation at a later date)?

Reply:
The idea would be to provide an opaque type. We can change it as long as we're in preview, but we need it to be stable; service owners can write tooling to use that ID. We should choose one and stick to it unless we really need to

Reply:
([source](https://datatracker.ietf.org/doc/html/draft-peabody-dispatch-new-uuid-format#page-11))

> UUIDv8's uniqueness will be implementation-specific and SHOULD NOT be assumed. [...] To be clear: UUIDv8 is not a replacement for UUIDv4 where all 122 extra bits are filled with random data.


([source](https://docs.rs/uuid/latest/uuid/)) 

> If you just want to generate unique identifiers then consider version 4 (v4) UUIDs. If you want to use UUIDs as database keys or need to sort them then consider version 7 (v7) UUIDs. Other versions should generally be avoided unless there’s an existing need for them.

We might not need this at the moment